lrauv_ignition_plugins/src/ScienceSensorsSystem.cc

```diff
@@ -53,13 +54,100 @@ class tethys::ScienceSensorsSystemPrivate
   /// \brief Generate octree from spatial data, for searching
   public: void GenerateOctrees();
 
+  /// \brief Interpolate among existing science data to output an estimated
+  /// reading at the current sensor location.
+  /// \param[in] _p Position to interpolate for
+  /// \param[in] _xyzs XYZ coordinates of existing data locations
+  /// \param[in] _values Data values at the locations
+  /// \return Interpolated value, or quiet NaN if inputs invalid.
+  public: float InterpolateData(
+    const Eigen::Vector3f &_p,
+    const Eigen::MatrixXf &_xyzs,
+    const std::vector<float> &_values);
+
+  /// \brief True to use trilinear interpolation, false to use barycentric
+  /// interpolation
+  public: bool useTrilinear = true;
+
   /// \brief Convert a vector of PCL points to an Eigen::Matrix.
   /// \param[in] _vec Source vector
   /// \param[out] _mat Result matrix
   public: void PclVectorToEigen(
     const std::vector<pcl::PointXYZ> &_vec,
     Eigen::MatrixXf &_mat);
 
+  //////////////////////////////////
+  // Trilinear interpolation
+
+  /// \brief Comparison function for std::set_difference().
+  /// Comparison between points is arbitrary. This function is only used for
+  /// set operations, not for literal sorting.
+  public: bool comparePclPoints(
+    const pcl::PointXYZ &a,
+    const pcl::PointXYZ &b);
+
+  /// \brief Find XYZ locations of points in the two closest z slices to
+  /// interpolate among.
+  /// \param[in] _pt Location in space to interpolate for
+  /// \param[in] _inds Indices of nearest neighbors to _pt
+  /// \param[in] _sqrDists Distances of nearest neighbors to _pt
+  /// \param[out] _interpolators1 XYZ points on a z slice to interpolate among
+  /// \param[out] _interpolators2 XYZ points on a second z slice to interpolate
+  /// among
+  public: void FindTrilinearInterpolators(
```

Comment:
Yeah, they should be cleared. feb33ed
We don't have standard success codes, at least in this project. I haven't seen them much in ign-gazebo plugins, because most plugin functions are in a private class internal to individual plugins. They may be called from only one place ever (as is the case here), so it doesn't need to be so formal.